python/cudf_polars/cudf_polars/experimental/expressions.py

```diff
@@ -0,0 +1,439 @@
+
+    Parameters
+    ----------
+    expr
+        The Expr node to decompose.
+    input_ir
+        The input IR node that ``expr`` will evaluate.
+    partition_info
+        A mapping from all unique IR nodes to the
+        associated partitioning information.
+    config_options
+        GPUEngine configuration options.
+    names
+        Generator of unique names for temporaries.
+
+    Returns
+    -------
+    expr
+        Decomposed Expr node.
+    input_ir
+        The rewritten ``input_ir`` to be evaluated by ``expr``.
+    partition_info
+        A mapping from unique nodes in the new graph to associated
+        partitioning information.
+    """
+    if isinstance(expr, Literal):
+        # For Literal nodes, we don't actually want an
+        # input IR with real columns, because it will
+        # mess up the result of ``HConcat``.
+        input_ir = Empty()
+        partition_info[input_ir] = PartitionInfo(count=1)
+
+    partition_count = partition_info[input_ir].count
+    if partition_count == 1 or expr.is_pointwise:
+        # Single-partition and pointwise expressions are always supported.
+        return expr, input_ir, partition_info
+    elif isinstance(expr, Agg) and expr.name in _SUPPORTED_AGGS:
+        # This is a supported Agg expression.
+        return _decompose_agg_node(
+            expr, input_ir, partition_info, config_options, names=names
+        )
+    else:
+        # This is an un-supported expression - raise.
+        raise NotImplementedError(
+            f"{type(expr)} not supported for multiple partitions."
+        )
+
+
+def _decompose(
+    expr: Expr, rec: ExprDecomposer
+) -> tuple[Expr, IR, MutableMapping[IR, PartitionInfo]]:
+    # Used by `decompose_expr_graph``
+
+    if not expr.children:
+        # Leaf node
+        return _decompose_expr_node(
+            expr,
+            rec.state["input_ir"],
+            {rec.state["input_ir"]: rec.state["input_partition_info"]},
+            rec.state["config_options"],
+            names=rec.state["unique_names"],
+        )
+
+    # Process child Exprs first
+    children, input_irs, _partition_info = zip(
+        *(rec(c) for c in expr.children), strict=True
+    )
+    partition_info = reduce(operator.or_, _partition_info)
+
+    # Assume the partition count is the maximum input-IR partition count
+    input_ir: IR
+    assert len(input_irs) > 0  # Must have at least one input IR
+    partition_count = max(partition_info[ir].count for ir in input_irs)
+    unique_input_irs = list(dict.fromkeys(input_irs))
+    if len(unique_input_irs) > 1:
+        # Need to make sure we only have a single input IR
+        # TODO: Check that we aren't concatenating misaligned
+        # columns that cannot be broadcasted. For example, what
+        # if one of the columns is sorted?
+        schema: MutableMapping[str, Any] = {}
+        for ir in unique_input_irs:
+            schema.update(ir.schema)
```

Comment:
*nit* (maybe a followup): should we check that none of the column names overlap?